### PR TITLE
Migrate google/best_practices.md to google/best_practices.ipynb as part of docs cleanup for Cirq 1.0

### PR DESCRIPTION
**Pull request by tanujkhattar**
- Fixes bugs in code demonstrating use of circuit operations and compiling them using new transformers. 

Follow up of https://github.com/quantumlib/Cirq/pull/5221

### CODE REVIEW

**Review by dabacon — APPROVED** [review]
LGTM

What does `deep=False` do?

File: docs/google/best_practices.md [hunk 1]
```diff
@@ -66,16 +66,16 @@ qubits = cirq.GridQubit.rect(2, 5)
 for i in range(100):
     my_circuit.append(cirq.H.on_each(*qubits))
     for q in qubits:
-        my_circuit.append(cirq.measure(q, key=f'm{q}'))
+        my_circuit.append(cirq.measure(q, key=cirq.MeasurementKey.parse_serialized(f'{i}:m{q}')))
```
Comment (95-martin-orion) [inline docs/google/best_practices.md:69]: Is `parse_serialized` necessary here? I'd prefer to keep the doc examples simple if possible.

File: docs/google/best_practices.md [hunk 2]
```diff
@@ -66,16 +66,16 @@ qubits = cirq.GridQubit.rect(2, 5)
...
 
 # The same circuit, but defined using CircuitOperations.
 # This is ~1000x smaller when serialized!
-sub_circuit = cirq.Circuit(cirq.H(qubits[0]), cirq.measure(qubits[0], key='m'))
-circuit_op = cirq.CircuitOperation(sub_circuit.freeze())
-circuit_op = circuit_op.with_qubits([q])
-circuit_op = circuit_op.with_measurement_key_mapping({'m': f'm{q}'})
-circuit_op = circuit_op.repeat(100)
-short_circuit = cirq.Circuit(circuit_op for q in qubits)
+q = cirq.NamedQubit("q")
```
Comment (95-martin-orion) [inline docs/google/best_practices.md:73]: This `q` should be renamed to avoid collision with `for q in qubits`.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Thanks for the catch, Tanuj! A couple of formatting requests.

File: docs/google/best_practices.md (hunk 1 above)
Comment (tanujkhattar) [reply docs/google/best_practices.md:69]: My reason to include parse_serialized was to ensure equivalence between original circuit and shortened circuit. IIUC, in the original circuit if we repeat the same measurement key 100 times, the sampled result will contain a 3d measurements table, whereas if each measurement key is unique (which is what happens if circuit operation is expanded) then the output format will be different. 

I'll be happy to change to whatever you recommend here.

**Review by MichaelBroughton — COMMENTED** [review]
Could we convert this to an ipynb file so the code snippets get tested in the docs infra ?

File: docs/google/best_practices.md (hunk 1 above)
Comment (95-martin-orion) [reply docs/google/best_practices.md:69]: Got it, that makes sense. Would `cirq.measure(q, key=f'{i}:m{q}')` in the original circuit work instead?

Comment (tanujkhattar) [reply docs/google/best_practices.md:69]: No, we cannot specify ':' character in the key string. 
```python
: is not allowed in MeasurementKey. If this is a nested key string, use `MeasurementKey.parse_serialized` for correct behavior.
```

File: docs/google/best_practices.md (hunk 2 above)
Comment (tanujkhattar) [reply docs/google/best_practices.md]: done

**Comment by tanujkhattar** [comment]
> What does `deep=False` do?

@dabacon `deep=False` will treat a circuit operation as a single top-level operation and not recursively optimize it. So, if we run a transformer to convert all single qubit gates to phxz, the transformer will look at a circuit operation on 2 qubits and simply ignore it as it's not a single qubit operation, instead of recursively running inside the circuit operation. Similarly, a transformer that merges 1 and 2 qubit connected components will compress an entire circuit operation and merge it's unitary matrix into operations to the left if the circuit operation is defined on <= 2 qubits and has a unitary.

File: docs/google/best_practices.md (hunk 1 above)
Comment (95-martin-orion) [reply docs/google/best_practices.md]: Yup, that's the bit I was missing. Thanks for confirming.

File: docs/tutorials/google/identifying_hardware_changes.ipynb [hunk 3]
```diff
@@ -720,7 +720,7 @@
     "\n",
     "You've selected better candidate qubits for your circuit, based on updated information about the device. What else can you do for further improvements? \n",
     "\n",
-    "*    You need to map your actual circuit's logical qubits to your selected hardware qubits. This is in general a difficult problem, and the best solution can depend on the specific structure of the circuit to be run. Take a look at the [Qubit Picking with Loschmidt Echoes](./echoes.ipynb) tutorial, which estimates the error rates of gates for your specific circuit. Also, consider [Best Practices#Qubit picking](../../google/best_practices.md#qubit-picking) for additional advice on this. \n",
+    "*    You need to map your actual circuit's logical qubits to your selected hardware qubits. This is in general a difficult problem, and the best solution can depend on the specific structure of the circuit to be run. Take a look at the [Qubit Picking with Loschmidt Echoes](./echoes.ipynb) tutorial, which estimates the error rates of gates for your specific circuit. Also, consider [Best Practices#Qubit picking](../../google/best_practices.ipynb#qubit-picking) for additional advice on this.\n",
```
Comment (95-martin-orion) [inline docs/tutorials/google/identifying_hardware_changes.ipynb:723]: ipynb header link doesn't seem to work for me (links to the page, but not the header)

**Review by 95-martin-orion — APPROVED** [review]
Just one nit, but I don't feel that it's blocking. Thanks for cleaning this up!

File: docs/tutorials/google/identifying_hardware_changes.ipynb (hunk 3 above)
Comment (tanujkhattar) [reply docs/tutorials/google/identifying_hardware_changes.ipynb]: Fixed the link so it works locally on the jupyter notebooks. I think the docs infra will replace a dash with an underscore so it also works on the website (based on my observation of existing links, where the link in the code uses a dash but if you click on the website, it is replaced with an underscore with works on the website). 

I'll watch out to verify once this is submitted.

**Pull request merged**